demo-orgs: Use same form flow when creating demo organizations.

laurynmm opened the pull request:
Currently, we've only been creating demo organizations via a button on the devlogin page. In preparation for implementing #34446, we need to update the normal new organization registration flow to handle the demo organization case.

**Notes**:
- Since folks creating a demo organization will need to agree to the "terms of service", we can still use the `RealmCreationForm` and `RegistrationForm` flow that's used in regular new organization creation.
  - Adds a `demo_organization` boolean field to the `PreregistrationRealm` model, so that we can track/set demo organization creation from the start of the process.
  -  When creating a demo organization, redirects from a valid `RealmCreationForm` to the registration page, bypassing the email confirmation step like we do for initial realm creation for setting up a new Zulip server.
- Because demo organizations do not need certain form inputs (email, subdomain, etc.), we adjust some of the field validation and cleaning to account for that case. As validation depends on two fields, moved some checks to happen at the form level (see https://docs.djangoproject.com/en/5.2/ref/forms/validation/#cleaning-and-validating-fields-that-depend-on-each-other).
- Currently, these changes don't include any updates to the templates and frontend code. The final commit is a small prep for passing a context variable for whether demo organizations are enabled.
  - I did light manual testing on the current new organization registration flow for normal organizations with these changes, e.g., making sure that errors for some of the validation/cleaning changes were not changed from the user-facing experience.

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on zerver/views/registration.py:
```diff
@@ -243,6 +246,21 @@ def get_selected_realm_default_language_name(
+            available_subdomain = True
+        except ValidationError:  # nocoverage
+            continue
+    return demo_subdomain
```
This is basically the same code that we've been using for demo organization subdomains in the dev environment, except in a while loop so that we definitely get an available subdomain.

Not sure if we want to either make this more or less random. I could imagine something a dashed string of random English words that starts with demo being perhaps nicer for users, e.g., random subdomains like "demo-apple-ladder-purple".

amanagr replied:
Yeah, we can use a package like https://github.com/the-lay/py-friendly-words for this. @alya FYI

amanagr commented on zerver/forms.py:
```diff
@@ -154,16 +154,34 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
             choices=[(lang["code"], lang["name"]) for lang in get_language_list()],
         )
 
+    @override
+    def clean(self) -> None:
+        super().clean()
+        if not self.realm_creation:
+            # The validation below only needs to happen when we are
+            # creating a new realm as part of the registration process.
+            return
+
+        create_demo = self.cleaned_data.get("create_demo", False)
+        if create_demo:
+            # Subdomains for demo organizations are not set by user input.
+            return
+
+        subdomain = self.cleaned_data.get("realm_subdomain", "")
+        try:
+            check_subdomain_available(subdomain)
+        except ValidationError as e:
+            self.add_error("realm_subdomain", e)
+
```
Is there a reason to separately override `clean` here instead of running the same logic in `clean_realm_subdomain`. Using `clean_realm_subdomain` seems preferable for this since the validation is related to it alone.

laurynmm replied:
Yes, the reason is that the validation/cleaning for `realm_subdomain` here does depend on the cleaned value of `create_demo`.

[Django docs for clean/validate form fields that depend on each other](https://docs.djangoproject.com/en/5.2/ref/forms/validation/#cleaning-and-validating-fields-that-depend-on-each-other).

amanagr commented on zerver/forms.py:
```diff
@@ -363,6 +365,19 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
+
+        email = self.cleaned_data.get("email", "")
+        if not email:
+            raise ValidationError(_("Please use your real email address."))
```
Should we move this logic check to `clean_email`?

laurynmm replied:
Whether we do this validation does depend on the `create_demo` field cleaned value. I used an early return instead of having both conditions, so this is really...

```python
if not create_demo and not email:
    raise ValidationError ...
```

[Django docs for clean/validate form fields that depend on each other](https://docs.djangoproject.com/en/5.2/ref/forms/validation/#cleaning-and-validating-fields-that-depend-on-each-other).

amanagr commented on zerver/tests/test_signup.py:
```diff
@@ -1857,6 +1854,16 @@ def test_create_education_demo_organization_welcome_bot_direct_message(self) ->
         self.assertIn("demo organization", welcome_msg.content)
 
+        # Organization owner has expected defaults for demo organization creation.
+        owner = realm.get_human_owner_users().last()
```
I think there is a possibility that this can return a different user, so we should use a different approach to test the created user.

laurynmm commented on zerver/forms.py:
```diff
@@ -236,12 +236,20 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
 
-    def clean_full_name(self) -> str:
+    @override
+    def clean(self) -> None:
```
Same reason as the other cases, the check for a valid `full_name` value depends on the cleaned value of `create_demo`.

Initially, I did try to modify the clean method for the specific fields, but it seemed odd that I needed to get the cleaned value for `create_demo` for that other field. The [Django docs ](https://docs.djangoproject.com/en/5.2/ref/forms/validation/#cleaning-and-validating-fields-that-depend-on-each-other)recommend using the form clean method when the value of one field depends on the value of another. From that link, with the bold text modified for our case:

> Suppose we add another requirement to our contact form: if the **create_demo field is False, the full_name field must be validated for various conditions (length, characters, if the existing realm allows duplicate names, etc.)**. We are performing validation on more than one field at a time, so the form’s [clean()](https://docs.djangoproject.com/en/5.2/ref/forms/api/#django.forms.Form.clean) method is a good spot to do this.

amanagr commented:
This needs a rebase for migration conflict, other than that this looks good to me.

laurynmm commented:
Rebased the changes here for the migration conflict.

For the changes to override the `clean` method on the forms, I think it's better to make it clearer that the various validation checks depend on the value of `create_demo`. So, I updated those changes to have comments explaining the need for the two fields in the validation check and removed the early return pattern so that the conditions were clearer when reading the code.

laurynmm commented:
**Updates**:
- Added commit to call the clean method for the superclass in the override clean method for `CaptchaRealmCreationForm`.

@alexmv mentioning you to make sure that the noted update above makes sense. It's the second commit in the series.

timabbott commented on zerver/forms.py:
```diff
@@ -139,16 +139,37 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
+        if not self.realm_creation:
+            # The validation below only needs to happen when we are
+            # creating a new realm as part of the registration process.
+            return
```
When is `realm_creation` expected to be false? Might be helpful for this comment to hint that.

timabbott commented on zerver/views/registration.py:
```diff
@@ -691,7 +691,10 @@ def registration_helper(
 
-        full_name = form.cleaned_data["full_name"]
+        if realm_creation and create_demo:
+            full_name = "Your name"
```
I think we may want generated default names to be something more like "Curious Crocodile" long term; but that can be discussed as a follow-up change.

alexmv commented:
> Added commit to call the clean method for the superclass in the override clean method for CaptchaRealmCreationForm.

Seems totally fine to me!

laurynmm commented on zerver/lib/test_classes.py:
```diff
@@ -956,14 +956,15 @@ def submit_realm_creation_form(
             "create_demo": create_demo,
         }
+        if not create_demo:
+            payload["email"] = email
```
So, the default empty value for the Django `EmailField` and `CharField` are an empty string. So not passing the value is probably what we want to be simulating in tests.

We could also choose to set the email and subdomain fields on the `PreregistrationRealm` object to the empty string value for the `create_demo` case, like we set the user full name later in the registration helper when creating the `User` object for the demo organization owner.

laurynmm commented:
**Updates**:
- Updated changes for review feedback above, and added individual notes to those comments.
- Will also post some additional comments about changes to the code that I think could use some feedback/input.

---

I'm building changes for https://github.com/zulip/zulip/issues/34446 on top of these updates, so we can hold off on merging these changes until the actual UI changes are in place for the registration form flow.

laurynmm commented:
Removing the integration review label as it looks like we're going in a design direction that will allow for two separate forms for demo and permanent organizations. See [#backend > demo organizations: registration forms @ 💬](https://chat.zulip.org/#narrow/channel/3-backend/topic/demo.20organizations.3A.20registration.20forms/near/2264642).
